File: common/service_response.go
package common

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// ServiceRequest an IoT service response
type ServiceResponse struct {
	// RequestId identity of the IoT service request to respond to
	RequestId string `json:"requestId"`

	// Time time of the IoT service response
	Time time.Time `json:"time"`

	// StatusCode status code of the IoT service response
	StatusCode int32 `json:"statusCode"`

	// ReturnValue return value of the IoT service response
	ReturnValue string `json:"returnValue"`
}

// GetKeyComponents return components that compose the IoT service response key
func (r *ServiceResponse) GetKeyComponents() []string {
	return []string{r.RequestId}
}

// Serialize transform current IoT service response to JSON string
func (r *ServiceResponse) Serialize() ([]byte, error) {
	return json.Marshal(r)
}

// Validate check if the IoT service response properties are valid
func (r *ServiceResponse) Validate() error {
	if _, err := uuid.Parse(r.RequestId); err != nil {
		return fmt.Errorf("invalid request ID in response definition")
	}
	if r.Time.IsZero() {
		return fmt.Errorf("missing response time in response definition")
	}

	return nil
}

// DeserializeService create an IoT service response instance from its JSON representation
func DeserializeServiceResponse(data []byte) (*ServiceResponse, error) {
	response := new(ServiceResponse)

	if err := json.Unmarshal(data, response); err != nil {
		return nil, err
	}

	return response, nil
}

File: sdk/javascript/ServiceResponse.ts
import { parse as uuidparse } from 'uuid';

import moment from './moment';

/**
 * An IoT service response
 */
export default class ServiceResponse {
  /**
   * @param requestId identity of the IoT service request to respond to
   * @param time time of the IoT service response
   * @param statusCode status code of the IoT service response
   * @param returnValue return value of the IoT service response
   */
  constructor(
    public requestId: string,
    public time: moment.Moment,
    public statusCode: number = 0,
    public returnValue: string = '',
  ) {}

  /**
   * Get components that compose the IoT service response key
   *
   * @returns components that compose the IoT service response key
   */
  getKeyComponents() {
    return [this.requestId];
  }

  /**
   * Transform current service response to a plain object
   *
   * @returns a plain object containing the service response information
   */
  toObject(): { [key: string]: any } {
    return {
      requestId: this.requestId,
      time: this.time,
      statusCode: this.statusCode,
      returnValue: this.returnValue,
    };
  }

  /**
   * Transform current IoT service response to JSON string
   *
   * @returns JSON representation of the service response
   */
  serialize() {
    return JSON.stringify(this.toObject());
  }

  /**
   * Check if the IoT service response properties are valid
   *
   * @throws error when service response fields are invalid
   */
  validate() {
    try {
      uuidparse(this.requestId);
    } catch {
      throw new Error('invalid request ID in response definition');
    }
    if (this.time.valueOf() === 0) {
      throw new Error('missing response time in response definition');
    }
  }

  /**
   * Create a new service response instance from an object
   *
   * @param obj an object that contains the service response information
   * @returns a new service response instance
   */
  static fromObject(obj: { [key: string]: any }): ServiceResponse {
    return new ServiceResponse(
      obj.requestId,
      moment(obj.time),
      obj.statusCode ? obj.statusCode : 0,
      obj.returnValue ? obj.returnValue : '',
    );
  }

  /**
   * Create an IoT service response instance from its JSON representation
   *
   * @param data JSON string representing a service response
   * @returns a new service response instance
   */
  static deserialize(data: string) {
    return ServiceResponse.fromObject(JSON.parse(data));
  }
}
